File: app/src/main/java/com/mecong/tenderalarm/alarm/AlarmNotifyingService.java
package com.mecong.tenderalarm.alarm;

import android.app.PendingIntent;
import android.app.Service;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.PowerManager;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.util.Log;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import com.hypertrack.hyperlog.HyperLog;
import com.mecong.tenderalarm.BuildConfig;
import com.mecong.tenderalarm.R;
import com.mecong.tenderalarm.model.AlarmEntity;
import com.mecong.tenderalarm.model.SQLiteDBHelper;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.io.IOException;
import java.util.Locale;
import java.util.Random;
import java.util.concurrent.TimeUnit;

import lombok.AccessLevel;
import lombok.experimental.FieldDefaults;
import timber.log.Timber;

import static androidx.core.app.NotificationCompat.FLAG_NO_CLEAR;
import static androidx.core.app.NotificationCompat.FLAG_ONGOING_EVENT;
import static androidx.core.app.NotificationCompat.FLAG_SHOW_LIGHTS;
import static com.mecong.tenderalarm.alarm.AlarmUtils.ALARM_ID_PARAM;
import static com.mecong.tenderalarm.alarm.AlarmUtils.TAG;

@FieldDefaults(level = AccessLevel.PRIVATE)
public class AlarmNotifyingService extends Service {
    private static final Uri CONTENT_URI = Uri.parse("content://" + BuildConfig.APPLICATION_ID + "/alarms");
    MediaPlayer alarmMediaPlayer;
    Handler handlerVolume;
    Runnable runnableVolume;
    Random random = new Random();
    Handler handlerTicks;
    Runnable runnableRealAlarm;
    MediaPlayer ticksMediaPlayer;

    @Override
    public void onCreate() {
        EventBus.getDefault().register(this);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String alarmId = intent.getStringExtra(ALARM_ID_PARAM);
        AlarmEntity entity = SQLiteDBHelper.getInstance(this).getAlarmById(alarmId);
        HyperLog.i(TAG, "Running alarm: " + entity);

        usePowerManagerWakeup();
        stopForeground(true);
        startAlarmNotification(this, entity);
        startSound(entity);

        return START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        // stopAlarmNotification();
    }

    private void usePowerManagerWakeup() {
        HyperLog.v(TAG, "Alarm Notifying service usePowerManagerWakeup");

        PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
        if (pm != null) {
            PowerManager.WakeLock wakeLock =
                    pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, this.getClass().getCanonicalName());
            wakeLock.acquire(TimeUnit.SECONDS.toMillis(10));
        }
    }

    @Subscribe
    public void messageReceived(AlarmMessage message) {
        if (message == AlarmMessage.CANCEL_VOLUME_INCREASE) {
            cancelVolumeIncreasing();
        } else if (message == AlarmMessage.STOP_ALARM) {
            stopAlarmNotification();
        } else if (message == AlarmMessage.SNOOZE2M) {
            snooze(2);
        } else if (message == AlarmMessage.SNOOZE3M) {
            snooze(3);
        } else if (message == AlarmMessage.SNOOZE5M) {
            snooze(5);
        }
    }

    private void snooze(int minutes) {
        HyperLog.d(TAG, "Snooze for " + minutes + " min");
        handlerTicks.removeCallbacksAndMessages(null);
        cancelVolumeIncreasing();
        if (ticksMediaPlayer.isPlaying()) {
            ticksMediaPlayer.pause();
            ticksMediaPlayer.seekTo(0);
        }
        if (alarmMediaPlayer.isPlaying()) {
            alarmMediaPlayer.pause();
            alarmMediaPlayer.seekTo(0);
        }
//        handlerTicks.postDelayed(runnableRealAlarm, minutes * 60 * 1000);
//        handlerVolume.postDelayed(runnableVolume, minutes * 60 * 1000);
    }

    private void cancelVolumeIncreasing() {
        handlerVolume.removeCallbacks(runnableVolume);
    }

    private void startSound(AlarmEntity entity) {

        final AudioAttributes audioAttributesAlarm = new AudioAttributes
                .Builder()
                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                .setUsage(AudioAttributes.USAGE_ALARM)
                .build();

        AudioManager audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        int streamMaxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_ALARM);
        audioManager.setStreamVolume(AudioManager.STREAM_ALARM, streamMaxVolume, 0);

        // Create the Handler object (on the main thread by default)
        handlerTicks = new Handler();
        final float[] volume = {0.001f};


        try {
            ticksMediaPlayer = new MediaPlayer();
            ticksMediaPlayer.setAudioAttributes(audioAttributesAlarm);
            ticksMediaPlayer.setDataSource(this, Uri.parse("android.resource://"
                    + getPackageName() + "/" + R.raw.tick));
            ticksMediaPlayer.prepare();
            ticksMediaPlayer.setVolume(volume[0], volume[0]);
        } catch (IOException e) {
            //TODO: make correct reaction
            e.printStackTrace();
        }

        try {
            alarmMediaPlayer = new MediaPlayer();
            alarmMediaPlayer.setAudioAttributes(audioAttributesAlarm);
            final Uri melody = getMelody(this, entity);
            alarmMediaPlayer.setDataSource(this, melody);
            alarmMediaPlayer.prepare();
            alarmMediaPlayer.setLooping(true);
        } catch (Exception ex) {
            Timber.e(ex);
        }

        final Runnable predAlarm = new Runnable() {
            @Override
            public void run() {
                try {
                    if (!ticksMediaPlayer.isPlaying()) {
                        ticksMediaPlayer.start();
                    }
                    HyperLog.v(TAG, "Tick!");
                } catch (Exception e) {
                    HyperLog.e(TAG, "Exception: " + e.getMessage(), e);
                }
                // Repeat this the same runnable code block again another 20 seconds
                // 'this' is referencing the Runnable object
                handlerTicks.postDelayed(this, random.nextInt(20000));
            }
        };


        runnableRealAlarm = new Runnable() {
            @Override
            public void run() {
                try {
                    // Removes pending code execution
                    handlerTicks.removeCallbacks(predAlarm);
                    if (ticksMediaPlayer.isPlaying()) {
                        ticksMediaPlayer.stop();
                    }

                    volume[0] = 0.01f;


                    alarmMediaPlayer.setVolume(volume[0], volume[0]);
                    alarmMediaPlayer.start();

                    handlerVolume.removeCallbacks(runnableVolume);
                    handlerVolume.post(runnableVolume);
                    HyperLog.i(TAG, "Real alarm started!");

                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };

        handlerVolume = new Handler();
        runnableVolume = new Runnable() {
            @Override
            public void run() {
                try {
                    volume[0] += 0.001f;
                    Log.v(TAG, "New alarm volume: " + volume[0]);
                    if (alarmMediaPlayer.isPlaying()) {
                        alarmMediaPlayer.setVolume(volume[0], volume[0]);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }

                if (volume[0] < 1)
                    handlerVolume.postDelayed(this, 2000);
            }
        };

        if (entity.getTicksTime() > 0) {
            handlerTicks.post(predAlarm);
        }

        handlerTicks.postDelayed(runnableRealAlarm, entity.getTicksTime() * 60 * 1000);
    }

    private Uri getMelody(Context context, AlarmEntity entity) {
        if (entity.getMelodyUrl() != null) {
            return Uri.parse(entity.getMelodyUrl());
        } else {
            return Uri.parse(String.format(Locale.ENGLISH, "android.resource://%s/%d",
                    context.getPackageName(), R.raw.long_music));
        }
    }

    private void stopAlarmNotification() {
        HyperLog.i(TAG, "Stop Alarm notification");
        handlerTicks.removeCallbacksAndMessages(null);
        stopTicksAlarm();
        stopAlarmMediaPlayer();
        alarmEndVibration();
        cancelVolumeIncreasing();
        stopForeground(true);
        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(this);
        notificationManager.cancelAll();
        stopSelf();
    }

    private void stopTicksAlarm() {
        if (ticksMediaPlayer.isPlaying()) {
            ticksMediaPlayer.stop();
            ticksMediaPlayer.reset();
            ticksMediaPlayer.release();
        }
    }

    private void stopAlarmMediaPlayer() {
        if (alarmMediaPlayer.isPlaying()) {
            alarmMediaPlayer.stop();
            alarmMediaPlayer.reset();
            alarmMediaPlayer.release();
        }
    }

    private void alarmEndVibration() {
        Vibrator vibrator = (Vibrator) getApplicationContext().getSystemService(Context.VIBRATOR_SERVICE);
        if (vibrator != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vibrator.vibrate(VibrationEffect.createOneShot(1000,
                        VibrationEffect.DEFAULT_AMPLITUDE));
            } else {
                vibrator.vibrate(1000);
            }
        }
    }


    private void startAlarmNotification(Context context, AlarmEntity entity) {
        String alarmId = String.valueOf(entity.getId());
        Intent startAlarmIntent = new Intent(context, AlarmReceiverActivity.class)
                .setData(ContentUris.withAppendedId(CONTENT_URI, Long.parseLong(alarmId)));
        startAlarmIntent.setFlags(
                Intent.FLAG_ACTIVITY_NEW_TASK
        );
        startAlarmIntent.putExtra(ALARM_ID_PARAM, alarmId);

        PendingIntent pendingIntent = PendingIntent.getActivity(context, 42,
                startAlarmIntent, PendingIntent.FLAG_UPDATE_CURRENT);


        NotificationCompat.Builder alarmNotification = new NotificationCompat
                .Builder(context, MainActivity.ALARM_CHANNEL_ID)
                .setSmallIcon(R.drawable.launcher)
                .setContentTitle("Alarm")
                .setContentText(entity.getMessage())
                .setContentIntent(pendingIntent)
                .setFullScreenIntent(pendingIntent, true)
                .setAutoCancel(false)
                .setOngoing(true)
                .setWhen(0)
                .setCategory(NotificationCompat.CATEGORY_ALARM)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setLocalOnly(false)
                .setDefaults(FLAG_SHOW_LIGHTS | FLAG_ONGOING_EVENT | FLAG_NO_CLEAR)
                .setPriority(NotificationCompat.PRIORITY_HIGH);

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
        notificationManager.cancelAll();

        startForeground(42, alarmNotification.build());
    }
}

File: app/src/main/java/com/mecong/tenderalarm/alarm/AlarmNotifyingService.kt
package com.mecong.tenderalarm.alarm

import android.app.AlarmManager
import android.app.PendingIntent
import android.app.Service
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.media.AudioManager
import android.net.Uri
import android.os.*
import android.os.Process.killProcess
import android.os.Process.myPid
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.C.CONTENT_TYPE_MUSIC
import com.google.android.exoplayer2.C.USAGE_ALARM
import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.upstream.RawResourceDataSource
import com.mecong.tenderalarm.BuildConfig
import com.mecong.tenderalarm.R
import com.mecong.tenderalarm.alarm.AlarmUtils.ALARM_ID_PARAM
import com.mecong.tenderalarm.alarm.AlarmUtils.ALARM_ID_PARAM_SAME_ID
import com.mecong.tenderalarm.model.AlarmEntity
import com.mecong.tenderalarm.model.SQLiteDBHelper.Companion.sqLiteDBHelper
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import timber.log.Timber
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit

class AlarmNotifyingService : Service(), Player.EventListener {
    private lateinit var runnableVolume: Runnable
    private lateinit var runnableRealAlarm: Runnable
    private lateinit var runnableVibro: Runnable

    private var random = Random()
    private var handlerVolume: Handler = Handler()
    private var handlerTicks: Handler = Handler()
    private var handlerVibration: Handler = Handler()
    private lateinit var exoPlayer: SimpleExoPlayer

    override fun onPlayerError(error: ExoPlaybackException) {
        super.onPlayerError(error)
        initExoPlayer(RawResourceDataSource.buildRawResourceUri(R.raw.dance_of_kaschey), Player.REPEAT_MODE_ONE)
        exoPlayer.playWhenReady = true
    }

    private fun initExoPlayer(streamUrl: Uri, repeatMode: Int) {
        exoPlayer = SimpleExoPlayer.Builder(applicationContext)
                .build()
                .apply {
                    setWakeMode(C.WAKE_MODE_LOCAL)
                    setHandleAudioBecomingNoisy(true)
                    addListener(this@AlarmNotifyingService)
                }

        val audioAttributesAlarm = com.google.android.exoplayer2.audio.AudioAttributes.Builder()
                .setContentType(CONTENT_TYPE_MUSIC)
                .setUsage(USAGE_ALARM)
                .build()

        exoPlayer.setAudioAttributes(audioAttributesAlarm, false)

        val dataSourceFactory = DefaultDataSourceFactory(this.baseContext, "Tender Alarm")
        val mediaSource: MediaSource = ProgressiveMediaSource.Factory(dataSourceFactory)
                .createMediaSource(streamUrl)
        exoPlayer.prepare(mediaSource)
        exoPlayer.repeatMode = repeatMode
    }

    override fun onCreate() {
        EventBus.getDefault().register(this)
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val action = intent.action
        if (action.equals(ACTION_STOP)) {
            stopAlarmNotification()
        } else {
            val alarmId = intent.getStringExtra(ALARM_ID_PARAM)
            val sameId = intent.getBooleanExtra(ALARM_ID_PARAM_SAME_ID, false)
            val entity = sqLiteDBHelper(this)!!.getAlarmById(alarmId)
            Timber.i("Alarm notification service start: $entity")
            usePowerManagerWakeup()

            if (sameId) {
                val pm = this.getSystemService(Context.POWER_SERVICE) as PowerManager
                val isScreenOn = pm.isInteractive
                Timber.i("SameId received; Notification activity shown: ${AlarmReceiverActivity.IS_SHOWN}; Screen on: $isScreenOn")
                if (!AlarmReceiverActivity.IS_SHOWN || !isScreenOn) {
                    startAlarmNotification(this, entity)
                }
                startSnoozedSound(entity)
            } else {
                startSound(entity)
                startAlarmNotification(this, entity)
            }
        }
        return START_NOT_STICKY
    }

    private fun usePowerManagerWakeup() {
        Timber.v("Alarm Notifying service usePowerManagerWakeup")
        val pm = getSystemService(Context.POWER_SERVICE) as PowerManager
        val wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, this.javaClass.canonicalName)
        wakeLock.acquire(TimeUnit.SECONDS.toMillis(10))
    }

    @Subscribe
    fun messageReceived(message: AlarmMessage) {
        when {
            message === AlarmMessage.CANCEL_VOLUME_INCREASE -> {
                cancelVolumeIncreasing()
            }
            message === AlarmMessage.STOP_ALARM -> {
                stopAlarmNotification()
            }
            message === AlarmMessage.SNOOZE2M -> {
                snooze(2)
            }
            message === AlarmMessage.SNOOZE3M -> {
                snooze(3)
            }
            message === AlarmMessage.SNOOZE5M -> {
                snooze(5)
            }
        }
    }

    private fun snooze(minutes: Int) {
        Timber.d("Snooze for $minutes min")
        handlerTicks.removeCallbacksAndMessages(null)
        cancelVolumeIncreasing()

        if (exoPlayer.isPlaying) {
            exoPlayer.stop()
            exoPlayer.seekTo(0)
        }
    }

    private fun cancelVolumeIncreasing() {
        handlerVolume.removeCallbacksAndMessages(null)
        handlerVibration.removeCallbacksAndMessages(null)
    }

    private fun startSnoozedSound(entity: AlarmEntity?) {
        handlerTicks.post(runnableRealAlarm)
    }

    private fun startSound(entity: AlarmEntity?) {
        val audioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val streamMaxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_ALARM)
        audioManager.setStreamVolume(AudioManager.STREAM_ALARM, streamMaxVolume, 0)

        val defaultVolume = 0.1f
        val volumeIncreaseDelayMs = 2000L
        val ticksTimeMs = entity!!.ticksTime * 60 * 1000
        val amountIncreases = ticksTimeMs / volumeIncreaseDelayMs
        var volumeIncreaseStep = if (amountIncreases.compareTo(0) == 0) 0.01f else (0.7f - defaultVolume) / amountIncreases

        val tickSound = if (entity.ticksType == 0) R.raw.tick else R.raw.ding1

        var volumeCounter = defaultVolume
        try {
            initExoPlayer(RawResourceDataSource.buildRawResourceUri(tickSound), Player.REPEAT_MODE_OFF)
            exoPlayer.volume = volumeCounter
        } catch (e: IOException) {
            //TODO: make correct reaction
            e.printStackTrace()
        }

        val predAlarm = object : Runnable {
            override fun run() {
                try {
                    if (!exoPlayer.isPlaying) {
                        exoPlayer.seekTo(0)
                        exoPlayer.playWhenReady = true
                    }

                    Timber.v("Tick!")
                } catch (e: Exception) {
                    //HyperLog.e(TAG, "Exception: " + e.message, e)
                }
                // Repeat this the same runnable code block again another 20 seconds
                // 'this' is referencing the Runnable object
                val twentySeconds = 20000
                handlerTicks.postDelayed(this, random.nextInt(twentySeconds).toLong())
            }
        }

        runnableRealAlarm = Runnable {
            try {
                // Removes pending code execution
                handlerTicks.removeCallbacksAndMessages(predAlarm)

                if (exoPlayer.isPlaying) {
                    exoPlayer.stop()
                    exoPlayer.release()
                }

                initExoPlayer(getMelody(entity), Player.REPEAT_MODE_ONE)

                exoPlayer.volume = if (entity.increaseVolume > 0) {
                    val increaseVolumeMinutes = entity.increaseVolume * 60000
                    volumeIncreaseStep = (1 - defaultVolume) / (increaseVolumeMinutes / volumeIncreaseDelayMs)
                    handlerVolume.removeCallbacks(runnableVolume)
                    handlerVolume.post(runnableVolume)
                    volumeCounter = defaultVolume
                    volumeCounter
                } else {
                    1f
                }

                exoPlayer.playWhenReady = true

                if (entity.vibrationType != null) {
                    handlerVibration.post(runnableVibro)
                }

                Timber.i("Real alarm started!")
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }


        runnableVolume = object : Runnable {
            override fun run() {
                try {
                    volumeCounter += volumeIncreaseStep
                    Timber.v("New alarm volume: $volumeCounter")

                    if (exoPlayer.isPlaying) {
                        exoPlayer.volume = volumeCounter
                    }

                } catch (e: Exception) {
                    e.printStackTrace()
                }
                if (volumeCounter < 1) handlerVolume.postDelayed(this, volumeIncreaseDelayMs)
            }
        }

        if (ticksTimeMs > 0) {
            handlerTicks.post(predAlarm)
            handlerVolume.post(runnableVolume)
        }

        handlerTicks.postDelayed(runnableRealAlarm, ticksTimeMs.toLong())

        runnableVibro = object : Runnable {
            override fun run() {
                alarmNotificationVibration(-1)
                handlerVibration.postDelayed(this, 10000)
            }
        }
    }

    private fun getMelody(entity: AlarmEntity?): Uri {
        return if (entity!!.melodyUrl != null) {
            Uri.parse(entity.melodyUrl)
        } else {
            RawResourceDataSource.buildRawResourceUri(R.raw.dance_of_kaschey)
        }
    }

    private fun stopAlarmNotification() {
        Timber.i("Stop Alarm notification")
        if (ALARM_PLAYING != null) {
            val entity = sqLiteDBHelper(this)!!.getAlarmById(ALARM_PLAYING)
            if (entity != null) {
                val alarmMgr = this.getSystemService(Context.ALARM_SERVICE) as AlarmManager
                AlarmUtils.turnOffSnoozeAlarm(entity, alarmMgr, this)
            }
        }

        handlerTicks.removeCallbacksAndMessages(null)
        stopAlarmSound()
        cancelVolumeIncreasing()
        alarmEndVibration()
        stopForeground(true)
        val notificationManager = NotificationManagerCompat.from(this)
        notificationManager.cancelAll()
        ALARM_PLAYING = null
        stopSelf()
        killProcess(myPid())
        EventBus.getDefault().unregister(this)
    }

    private fun stopAlarmSound() {
        if (exoPlayer.isPlaying) {
            exoPlayer.stop()
            exoPlayer.release()
        }
    }

    private fun alarmEndVibration() {
        val vibrator = applicationContext.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(1000,
                    VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            vibrator.vibrate(1000)
        }
    }

    private fun alarmNotificationVibration(repeat: Int) {
        val vibrator = applicationContext.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        val mVibratePattern = longArrayOf(400, 400, 1000, 600, 600, 800, 500, 500)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createWaveform(mVibratePattern, repeat))
        } else {
            vibrator.vibrate(mVibratePattern, repeat)
        }
    }

    private fun startAlarmNotification(context: Context, entity: AlarmEntity?) {
        val alarmId = entity!!.id.toString()

        val stopIntent = Intent(context, AlarmNotifyingService::class.java)
        stopIntent.action = ACTION_STOP
        val stopAction = PendingIntent.getService(context, 3, stopIntent, 0)

        Timber.i("Start alarm notification: %s", entity)

        val startAlarmIntent = Intent(context, AlarmReceiverActivity::class.java)
                .setData(ContentUris.withAppendedId(CONTENT_URI, alarmId.toLong()))
        startAlarmIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        startAlarmIntent.putExtra(ALARM_ID_PARAM, alarmId)

        val pendingIntent = PendingIntent.getActivity(context, entity.fullScreenIntentCode,
                startAlarmIntent, PendingIntent.FLAG_CANCEL_CURRENT)

        val alarmNotification = NotificationCompat.Builder(context, MainActivity.ALARM_CHANNEL_ID)
                .setSmallIcon(R.drawable.cat_sleep)
                .setContentTitle(context.getString(R.string.alarm_ringing))
                .setContentText(entity.message)
                .setContentIntent(pendingIntent)
                .setFullScreenIntent(pendingIntent, true)
                .setLargeIcon(BitmapFactory.decodeResource(resources, R.drawable.cat_banjo))
                .setSmallIcon(R.drawable.alarm_active)
                .setAutoCancel(false)
                .setOngoing(true)
                .setWhen(0)
                .addAction(android.R.drawable.ic_menu_close_clear_cancel, context.getString(R.string.stop), stopAction)
                .setCategory(NotificationCompat.CATEGORY_ALARM)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setLocalOnly(false)
                .setDefaults(NotificationCompat.FLAG_SHOW_LIGHTS or NotificationCompat.FLAG_ONGOING_EVENT or NotificationCompat.FLAG_NO_CLEAR)
                .setPriority(NotificationCompat.PRIORITY_HIGH)

        val notificationManager = NotificationManagerCompat.from(context)
        notificationManager.cancelAll()
        startForeground(43, alarmNotification.build())
    }


    companion object {
        const val ACTION_STOP = "com.mecong.myalarm.ACTION_STOP"
        var ALARM_PLAYING: String? = null
        private val CONTENT_URI = Uri.parse("content://" + BuildConfig.APPLICATION_ID + "/alarms")
    }
}